src/backend/commands/copy.c

```diff
@@ -722,32 +723,84 @@ CopyGetData(CopyState cstate, void *databuf, int datasize)
 	switch (cstate->copy_dest)
 	{
 		case COPY_FILE:
-			bytesread = fread(databuf, 1, datasize, cstate->copy_file);
-			if (feof(cstate->copy_file))
-				cstate->fe_eof = true;
-			if (ferror(cstate->copy_file))
+			if (cstate->is_program)
 			{
-				if (cstate->is_program)
+				struct pollfd pollfd;
+				int			rc;
+				ssize_t		result = 0;
+
+				/* Check if data is available */
+				pollfd.fd = fileno(cstate->copy_file);
+				pollfd.events = POLLIN;
+				pollfd.revents = 0;
+				rc = poll(&pollfd, 1, 0);
+
+				/*
+				 * Continue waiting until we get data, even if interrupted by
+				 * a signal
+				 */
+				while (rc == 0 || (rc < 0 && (errno == EINTR || errno == EAGAIN)))
```

Comment:
A do..while loop would force `CHECK_FOR_INTERRUPTS` to always happen at least once on every call. Since there were no interrupt checks in the original code, I thought I shouldn't introduce them in case they are not necessary (like with fast connections that always have data available without waiting). I could also move `poll` before `CHECK_FOR_INTERRUPTS` inside the loop and exit it with a `break`, but then this wastes time waiting on `poll` when it could be checking for interrupts already. Should I do that anyway?

Reply:
CHECK_FOR_INTERRUPTS() is fast, no problem with call it one more time.

Reply:
Okay, then I will simplify this